pkg/uprobetracer/tracer.go

```diff
@@ -0,0 +1,245 @@
+)
+
+type inodeKeeper struct {
+	// if an inodeKeeper exists, `file` must be available, but `link` may be nil
+	// so if a inode is proved to not having right symbol, we can skip it for the next time
+	counter int
+	file    *os.File
+	link    link.Link
+}
+
+func (t *inodeKeeper) close() {
+	t.file.Close()
+	if t.link != nil {
+		t.link.Close()
+	}
+}
+
+type Tracer[Event any] struct {
+	progName       string
+	progType       string
+	attachFilePath string
+	attachSymbol   string
+	prog           *ebpf.Program
+
+	ctrPid2FsId    map[uint32]unix.Fsid
+	inodeRefCount  map[unix.Fsid]*inodeKeeper
+	pendingCtrPids map[uint32]bool
+
+	closed bool
+
+	mu sync.Mutex
+}
+
+func NewTracer[Event any]() (_ *Tracer[Event], err error) {
+	t := &Tracer[Event]{
+		ctrPid2FsId:    make(map[uint32]unix.Fsid),
+		inodeRefCount:  make(map[unix.Fsid]*inodeKeeper),
+		pendingCtrPids: make(map[uint32]bool),
+		closed:         false,
+	}
+	return t, nil
+}
+
+func (t *Tracer[Event]) attachProgToInode(file *os.File) (link.Link, error) {
+	// attach ebpf program to self-hosted inode, used by `Attach()`
+
+	attachPath := path.Join(host.HostProcFs, "self/fd/", fmt.Sprint(file.Fd()))
+	ex, err := link.OpenExecutable(attachPath)
+	if err != nil {
+		return nil, fmt.Errorf("opening executable: %q", attachPath)
+	}
+	switch t.progType {
+	case "uprobe":
+		return ex.Uprobe(t.attachSymbol, t.prog, nil)
+	case "uretprobe":
+		return ex.Uretprobe(t.attachSymbol, t.prog, nil)
+	default:
+		return nil, fmt.Errorf("attaching to inode: unsupported prog type: %q", t.progType)
+	}
+}
+
+func (t *Tracer[Event]) AttachProg(progName string, progType string, attachTo string, prog *ebpf.Program) error {
+	// interface for outside, load ebpf program, and attach if there are pending containers
+
+	if prog == nil {
+		return fmt.Errorf("prog does not exist")
+	}
+	if t.prog != nil {
+		return fmt.Errorf("loading uprobe program twice")
+	}
+
+	parts := strings.Split(attachTo, ":")
+	if len(parts) < 2 {
+		return fmt.Errorf("invalid section name %q", attachTo)
+	}
+	if !filepath.IsAbs(parts[0]) && strings.Contains(parts[0], "/") {
+		return fmt.Errorf("section name must be either an absolute path or a library name: %q", parts[0])
+	}
+	if progType != "uprobe" && progType != "uretprobe" {
+		return fmt.Errorf("unsupported uprobe prog type: %q", progType)
+	}
+	if strings.Contains(parts[0], "..") {
+		return fmt.Errorf("'..' is not allowed to appear in the path")
+	}
+
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	t.progName = progName
+	t.progType = progType
+	t.attachFilePath = parts[0]
+	t.attachSymbol = parts[1]
+	t.prog = prog
+
+	// attach to pending inode, then release the pending list
+	for pid := range t.pendingCtrPids {
+		t.attach(pid)
+	}
+	t.pendingCtrPids = nil
+
+	return nil
+}
+
+func searchForLibrary(containerPid uint32, path string) (string, error) {
+	if filepath.IsAbs(path) {
+		return filepath.Join(host.HostProcFs, fmt.Sprint(containerPid), "root", path), nil
```

Comment:
Consider using https://github.com/cyphar/filepath-securejoin

Credits to @rata for the link :)

```golang
SecureJoin(filepath.Join(host.HostProcFs, fmt.Sprint(containerPid), "root"), path)
```